probe-rs-cli-util/src/common_options.rs

```diff
@@ -0,0 +1,453 @@
+pub struct CargoOptions {
+    #[structopt(name = "binary", long = "bin", hidden = true)]
+    pub bin: Option<String>,
+    #[structopt(name = "example", long = "example", hidden = true)]
+    pub example: Option<String>,
+    #[structopt(name = "package", short = "p", long = "package", hidden = true)]
+    pub package: Option<String>,
+    #[structopt(name = "release", long = "release", hidden = true)]
+    pub release: bool,
+    #[structopt(name = "target", long = "target", hidden = true)]
+    pub target: Option<String>,
+    #[structopt(name = "PATH", long = "manifest-path", hidden = true)]
+    pub manifest_path: Option<String>,
+    #[structopt(long, hidden = true)]
+    pub no_default_features: bool,
+    #[structopt(long, hidden = true)]
+    pub all_features: bool,
+    #[structopt(long, hidden = true)]
+    pub features: Vec<String>,
+}
+
+impl CargoOptions {
+    pub fn help_message(bin: &str) -> String {
+        format!(
+            r#"
+CARGO BUILD OPTIONS:
+
+    The following options are forwarded to 'cargo build':
+
+        --bin
+        --example
+    -p, --package
+        --release
+        --target
+        --manifest-path
+        --no-default-features
+        --all-features
+        --features
+
+    For example, if you run the command '{} --release',
+    this means that 'cargo build --release' will be called.
+"#,
+            bin
+        )
+    }
+
+    pub fn to_cargo_arguments(&self) -> Vec<String> {
+        let mut args: Vec<String> = vec![];
+        macro_rules! maybe_push_str_opt {
+            ($field:expr, $name:expr) => {{
+                if let Some(value) = $field {
+                    args.push(format!("--{}", stringify!($name)));
+                    args.push(value.clone());
+                }
+            }};
+        }
+
+        maybe_push_str_opt!(&self.bin, bin);
+        maybe_push_str_opt!(&self.example, example);
+        maybe_push_str_opt!(&self.package, package);
+        if self.release {
+            args.push("--release".to_string());
+        }
+        maybe_push_str_opt!(&self.bin, bin);
+        #[rustfmt::skip]
+        maybe_push_str_opt!(&self.manifest_path, manifest-path);
+        if self.no_default_features {
+            args.push("--no-default-features".to_string());
+        }
+        if self.all_features {
+            args.push("--all-features".to_string());
+        }
+        if !self.features.is_empty() {
+            args.push("--features".to_string());
+            args.push(self.features.join(","));
+        }
+
+        args
+    }
+}
+
+#[derive(Debug, thiserror::Error)]
+pub enum OperationError {
+    #[error("No connected probes were found.")]
+    NoProbesFound,
+    #[error("Failed to list the target descriptions.")]
+    FailedToReadFamilies(#[source] RegistryError),
+    #[error("Failed to open the ELF file '{path}' for flashing.")]
+    FailedToOpenElf {
+        #[source]
+        source: std::io::Error,
+        path: String,
```

Comment:
Paths are usually stored as a `PathBuf`. What's the reason for using a `String` here? (Ditto for the other variants).

Reply:
Copied verbatim from `cargo-flash`. I did one `PathBuf` -> `String` substitution when writing `to_cargo_arguments`, but it was needless. I'll give the options a once-over.